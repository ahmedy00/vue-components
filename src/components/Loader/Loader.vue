<script setup lang="ts">
import { computed } from 'vue'

export type Props = {
  show: boolean,
  size: 'small' | 'medium' | 'large'
  text: string
  theme: 'light' | 'dark'
}

const props = withDefaults(defineProps<Props>(), {
  show: false,
  size: 'medium',
  text: '',
  theme: 'light'
})

// TODO: Will be added px option
const spinnerSize = computed(() => {
  const sizes = {
    small: 10,
    medium: 12,
    large: 16
  }
  return sizes[props.size]
})

const textColor = computed(() => {
  return props.theme === 'light' ? 'black' : 'white'
})

</script>

<template>
  <div
    data-test="loader"
    v-if="show"
    class="flex flex-col items-center justify-center space-y-1"
  >
    <div
      data-test="spinner"
      class="border-4 border-t-4 border-t-primary rounded-full animate-spin ease-linear"
      :class="`w-${spinnerSize} h-${spinnerSize}`"
    />
    <div
      data-test="loader-text"
      v-if="!!text"
      :class="`text-${textColor}`"
    >
      {{ text }}
    </div>
  </div>
</template>