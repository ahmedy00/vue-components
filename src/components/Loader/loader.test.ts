import { beforeEach, describe, expect, test } from 'vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import Loader from '@/components/Loader/Loader.vue'

describe('Loader Component', () => {
  let wrapper: VueWrapper<any>

  beforeEach(() => {
    wrapper = mount(Loader, {
      props: {
        show: true,
        size: 'large',
        text: 'Please Wait',
        theme: 'dark'
      }
    })
  })

  describe('Fundamental Tests', () => {
    test('renders component when `show` prop is true', () => {
      expect(wrapper.find('[data-test="loader"]').exists()).toBe(true)
    })

    test('renders text when text prop is exists', () => {
      expect(wrapper.find('[data-test="loader-text"]').exists()).toBe(true)
    })
  })

  describe('Style Tests', () => {
    test('has `w-16` and `h-16` classes while size is large', () => {
      console.log('--------------------------------classList------------------------', wrapper.find('[data-test="spinner"]').element.classList)
      expect(wrapper.find('[data-test="spinner"]').classes()).toContain('w-16')
      expect(wrapper.find('[data-test="spinner"]').classes()).toContain('h-16')
    })

    test('has `text-white` class while theme is dark', () => {
      expect(wrapper.find('[data-test="loader-text"]').classes()).toContain('text-white')
    })

    test('has correct default styles', () => {
      expect(wrapper.find('[data-test="loader"]').element).toMatchSnapshot()
    })
  })
})