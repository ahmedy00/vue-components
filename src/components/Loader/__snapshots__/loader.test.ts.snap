// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Loader Component > Style Tests > has correct default styles 1`] = `
<div
  class="flex flex-col items-center justify-center space-y-1"
  data-test="loader"
>
  <div
    class="border-4 border-t-4 border-t-primary rounded-full animate-spin ease-linear w-16 h-16"
    data-test="spinner"
  />
  <div
    class="text-white"
    data-test="loader-text"
  >
    Please Wait
  </div>
</div>
`;
